### PR TITLE
Automatically save game before exit

### PR DESCRIPTION
**Pull request by o01eg**
Save game in hostless mode if server stops playing the game.
Prerequisite for last step of #2243 so partial save will be available.

### CODE REVIEW

File: server/ServerFSM.cpp [hunk 1]
```diff
@@ -473,6 +492,18 @@ bool ServerFSM::EstablishPlayer(const PlayerConnectionPtr& player_connection,
...
 }
 
+std::string ServerFSM::GetAutoSaveFileName() {
+    std::string subdir = GetGalaxySetupData().GetGameUID();
+    boost::filesystem::path autosave_dir_path = GetServerSaveDir() / (subdir.empty() ? "auto" : subdir);
+    const auto& extension = MP_SAVE_FILE_EXTENSION;
+    // Add timestamp to autosave generated files
+    std::string datetime_str = FilenameTimestamp();
+
+    std::string save_filename = boost::io::str(boost::format("FreeOrion_%04d_%s%s") % m_server.CurrentTurn() % datetime_str % extension);
```
Comment (geoffthemedio) [inline server/ServerFSM.cpp]: is accessing `m_server.CurrentTurn()` the only reason this function is a member of `ServerFSM`?

Comment (o01eg) [reply server/ServerFSM.cpp]: No, this code to generate save file name is used in two places now so it is also refactored to reduce code repetition. `ServerFSM` is a common space for both usages.

Comment (geoffthemedio) [reply server/ServerFSM.cpp]: my point was whether that was the only reason for it not to be a file-scope free function that doesn't appear in the header

Comment (o01eg) [reply server/ServerFSM.cpp]: Fixed.

**Pull request merged**